Adding Firebase crashlytics

langsmith opened the pull request:
Resolves #625

langsmith commented:
Phew, finally ready for review when ya'll get a chance.

It's no longer in the actual pr, but during testing, I set up forcing a test crash as documented at https://firebase.google.com/docs/crashlytics/force-a-crash.

![ezgif com-resize](https://user-images.githubusercontent.com/4394910/60140855-eae7db80-9767-11e9-9150-babfefa1ce89.gif)

After linking the debug version from Fabric to Firebase, the crashes started to come into the Firebase console.

![Screen Shot 2019-07-03 at 1 29 48 PM](https://user-images.githubusercontent.com/4394910/60623080-a9be7f80-9d96-11e9-877b-bd55c4fdb0be.png)

![Screen Shot 2019-07-03 at 1 30 40 PM](https://user-images.githubusercontent.com/4394910/60623136-c8bd1180-9d96-11e9-9c85-3fd88265dc60.png)

Couple of things to know about this pr:

1. Dependency management got kinda' wild while dealing with new/upgraded Firebase libraries. As the error messages continued one after another, it became clear that migrating to AndroidX dependencies would be the best choice for dealing with these conflicts and for the app in general moving forward. Thanks in large part to Android Studio's automatic AndroidX migration tool, this pr transitions the app to AndroidX. Most of this pr's changed classes are because of importing new AndroidX packages. Use the file filter if you want to a see a more manageable overview of this pr's changes (https://github.com/mapbox/mapbox-android-demo/pull/1104/files?file-filters%5B%5D=.gradle&file-filters%5B%5D=.yml&file-filters%5B%5D=dotfile excludes .java/.kt/XML files)
2. The Firebase setup required me to bump the `minSDK` to 16. I became comfortable with this new minSDK requirement after looking at the `Platform versions` breakdown on https://developer.android.com/about/dashboards. I'm comfortable with excluding the .6% of devices that are pre-16. If you have any objections to bumping this, please let me know.
3. The `initializeFirebaseApp()` that was in the `MapboxApplication` class, was removed because the initialization is now done behind the scenes by Firebase's core library.
4. Firebase requires the Google Services plugin and `google-services.json` file in the `MapboxAndroidDemo` folder to do the crash reporting. The JSON file holds special keys, so it can't be publically committed to Github. I've added the release flavor's Firebase JSON file text as a CircleCI environment variable and created a `circle.yml` task to create the file during CI runs.

tobrun commented on MapboxAndroidDemo/src/global/java/com/mapbox/mapboxandroiddemo/MapboxApplication.java:
```diff
@@ -46,4 +38,12 @@ private void setUpTileLoadingMeasurement() {
+  private void setUpCrashlytics() {
+    Fabric.with(this, new Crashlytics.Builder()
+        .core(new CrashlyticsCore.Builder()
+            .disabled(BuildConfig.DEBUG)
```
We are configuring not collecting any data for debug builds in https://github.com/mapbox/mapbox-android-demo/pull/1104/files#diff-5f14090300b04edf8f7f1a0190d2872aR76. Do we need to keep both configurations?

langsmith replied:
Yes, seems that it's best to have both configurations 

- https://docs.fabric.io/android/crashlytics/build-tools.html#disable-crashlytics-for-debug-builds
- https://stackoverflow.com/a/28357998/6358488
- https://stackoverflow.com/a/16990788/6358488

tobrun commented on circle.yml:
```diff
@@ -33,13 +33,33 @@ jobs:
+            echo "android.useAndroidX=true" >> gradle.properties
+            echo "android.enableJetifier=true" >> gradle.properties
+            echo "org.gradle.jvmargs=-Xmx4608M" >> gradle.properties
+            cat gradle.properties
```
can be removed

langsmith replied:
Removed

tobrun commented on circle.yml:
```diff
@@ -33,13 +33,33 @@ jobs:
+            rm -fr MapboxAndroidDemo/fabric.properties
+            echo "apiKey=${FABRIC_API_KEY}" >> MapboxAndroidDemo/fabric.properties
+            echo "apiSecret=${FABRIC_API_SECRET}" >> MapboxAndroidDemo/fabric.properties
+            cat MapboxAndroidDemo/fabric.properties
```
we shouldn't print secret env variables to the console

langsmith replied:
Removed

tobrun commented on circle.yml:
```diff
@@ -33,13 +33,33 @@ jobs:
+      - run:
+          name: Add preferences to fabric.properties file
+          command: |
+            rm -fr MapboxAndroidDemo/fabric.properties
```
every docker starts from a fresh checkout. I believe we don't need to keep this `rm -fr` around

langsmith replied:
Removed

langsmith commented:
Ok @tobrun , I've addressed your feedback

> Main request for change is to optimize circle configuration for external pull request. We don't allow using env variables for them and we shouldn't execute steps that rely on them. See https://github.com/mapbox/mapbox-gl-native/blob/master/circle.yml#L573 for an example.

I believe https://github.com/mapbox/mapbox-android-demo/pull/1104/commits/6e313e05b6f5ecbd78aa3104cc79c10035beae56 correctly addresses the key/secret generation in the way you were envisioning. If not, let me know!

tobrun commented on circle.yml:
```diff
@@ -33,13 +33,31 @@ jobs:
+          name: Create google-services.json file
+          shell: /bin/bash -euo pipefail
+          command: |
+            echo "${GOOGLE_SERVICES_JSON_RELEASE}" > MapboxAndroidDemo/google-services.json
```
this need the same configuration as FABRIC_PROPERTIES_CONFIG_FILE below

langsmith replied:
Adjusted

tobrun commented on circle.yml:
```diff
@@ -33,13 +33,31 @@ jobs:
+      - run:
+          name: Add preferences to gradle.properties file
+          command: |
+            rm -fr gradle.properties
```
why are removing this file, afaik it won't exist when doing a checkout?

langsmith replied:
Left over from other experimentation. Removed.
